File: problems/1371_find_the_longest_substring_containing_vowels_in_even_counts/solution.py
class Solution:
    def findTheLongestSubstring(self, s: str) -> int:
        vowels = {'a': 0, 'e': 1, 'i': 2, 'o': 3, 'u': 4}
        state = 0
        first = {0: -1}
        res = 0
        for i, c in enumerate(s):
            if c in vowels:
                state ^= (1 << vowels[c])
            if state in first:
                res = max(res, i - first[state])
            else:
                first[state] = i
        return res

File: problems/1371_find_the_longest_substring_containing_vowels_in_even_counts/solution.cpp
#include <vector>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>
#include <climits>
#include <queue>
#include <stack>
#include <cmath>
#include <numeric>
#include <set>
#include <map>
using namespace std;

class Solution {
public:
    int findTheLongestSubstring(string s) {
        unordered_map<int,int> first;
        first[0] = -1;
        int state = 0, res = 0;
        string vowels = "aeiou";
        for (int i = 0; i < (int)s.size(); i++) {
            int idx = vowels.find(s[i]);
            if (idx != (int)string::npos) state ^= (1 << idx);
            if (first.count(state)) res = max(res, i - first[state]);
            else first[state] = i;
        }
        return res;
    }
};
